tools/benchmark/cmd/watch_latency.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	v3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/pkg/report"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+	"gopkg.in/cheggaaa/pb.v1"
+)
+
+// watchLatencyCmd represents the watch latency command
+var watchLatencyCmd = &cobra.Command{
+	Use:   "watch-latency",
+	Short: "Benchmark watch latency",
+	Long: `Benchmark Latency watch tests the latency of 
+	watch response processing by sending events to the watch. 
+	It tests the watch latency by changing the value of the 
+	keys with concurrent put requests and finding the latency 
+	time for the watcher.`,
+	Run: watchLatencyFunc,
+}
+
+var (
+	watchLValueTotal   int
+	watchLPutRate      int
+	watchLKeySize      int
+	watchLValueSize    int
+	watchLValSpaceSize int
+	watchLSeqVals      bool
+)
+
+func init() {
+	RootCmd.AddCommand(watchLatencyCmd)
+	watchLatencyCmd.Flags().IntVar(&watchLValueTotal, "watch-value-for-key", 10000, "Total number of values to watch for a key")
+	watchLatencyCmd.Flags().IntVar(&watchLPutRate, "put-rate", 100, "Number of keys to put per second")
+	watchLatencyCmd.Flags().IntVar(&watchLKeySize, "key-size", 32, "Key size of watch request")
+	watchLatencyCmd.Flags().IntVar(&watchLValueSize, "val-size", 32, "Val size of watch request")
+	watchLatencyCmd.Flags().IntVar(&watchLValSpaceSize, "val-space-size", 1, "Maximum possible values")
+	watchLatencyCmd.Flags().BoolVar(&watchLSeqVals, "sequential-values", false, "Use sequential values")
+}
+
+func watchLatencyFunc(cmd *cobra.Command, args []string) {
+	if watchLValSpaceSize <= 0 {
+		fmt.Fprintf(os.Stderr, "expected positive --key-space-size, got (%v)", watchLValSpaceSize)
+		os.Exit(1)
+	}
+
+	key := string(mustRandBytes(watchLKeySize))
+
+	values := make([]string, watchLValueTotal)
```

Comment:
This should be keys and use something like `watchLKeyTotal`. The total number of rounds should be unrelated to the number of keys being updated (the watch command makes a distinction between total keys and total received requests), otherwise there's no way to run the benchmark for a long time without exhausting memory.